### PR TITLE
Added support for `pnpm` in CKEditor 5 package generator.

### LINKED ISSUE #247: Add support for pnpm@10 in CKEditor 5 package generator
## 📝 Provide a description of the new feature

The CKEditor 5 package generator currently supports preparing the development environment using `npm` or `yarn` as the package manager. With the recent adoption and popularity of pnpm, we should extend support to this package manager to make onboarding smoother for developers who prefer it.

### Definition of Done

- Developers can successfully generate a CKEditor 5 package using pnpm@10.
- All scripts (build, test, lint, etc.) work out of the box with pnpm.
- CI passes using pnpm as the package manager.
- Documentation clearly shows pnpm as a supported option alongside npm and yarn.

### Other details

Let's keep `yarn` as the default so far.

### PR DESCRIPTION
**Pull request by martnpaneq**
### 🚀 Summary

Added comprehensive support for `pnpm` alongside existing `npm` and `yarn` options in the CKEditor 5 package generator.

Features added:

* New `--use-pnpm` CLI flag for specifying pnpm as package manager.
* Using `pnpm@10` on CI.

---

### 📌 Related issues

* Closes #247 

---

### 💡 Additional information

*Optional: Notes on decisions, edge cases, or anything helpful for reviewers.*

### CODE REVIEW

**Comment by martnpaneq** [comment]
How to test the change:

1. Testing that `pnpm` as a tool to install dependencies works.
    - It works on the CI. https://app.circleci.com/pipelines/github/ckeditor/ckeditor5-package-generator/1056/workflows/91600702-eb9d-41f3-ad87-ca9305682fd1/jobs/7762
    - You can test that basic commands like: `pnpm test`, `pnpm lint` work fine.
 2. Testing if `pnpm` as an argument to the package generator works.
    - It works on the CI: https://app.circleci.com/pipelines/github/ckeditor/ckeditor5-package-generator/1056/workflows/91600702-eb9d-41f3-ad87-ca9305682fd1/jobs/7764
    - To test it locally run: `node scripts/ci/verify-build -l js -p pnpm -m current `.
    - Test with other variants, e.g. `-l js`, '-m current-and-legacy`.

File: packages/ckeditor5-package-generator/lib/utils/choose-package-manager.js [hunk 1]
```diff
@@ -5,24 +5,27 @@
...
  * @param {Boolean} useNpm
  * @param {Boolean} useYarn
- * @returns {Promise<'npm'|'yarn'>}
+ * @param {Boolean} usePnpm
+ * @returns {Promise<'npm'|'yarn'|'pnpm'>}
  */
-export default async function choosePackageManager( useNpm, useYarn ) {
+export default async function choosePackageManager( useNpm, useYarn, usePnpm ) {
 	const yarnInstalled = isYarnInstalled();
+	const pnpmInstalled = isPnpmInstalled();
```
Comment (martnpaneq) [inline packages/ckeditor5-package-generator/lib/utils/choose-package-manager.js:18]: I'm thinking whether we should remove checking if `yarn` or `pnpm` is installed. If they are not, an error would be thrown anyway, and it should be pretty obvious that they should be installed.

We could remove two functions: `isYarnInstalled`, `isPnpmInstalled`, and simplify the logic.

File: .npmrc [hunk 2]
```diff
@@ -0,0 +1,2 @@
+frozen-lockfile=true
+node-linker=hoisted
```
Comment (filipsobol) [inline .npmrc:2]: Instead of storing this configuration in a separate file, add the following options to `pnpm-workspace.yaml`:

```
shamefullyHoist: true
preferFrozenLockfile: true
```

You can use the following PR as a reference: https://github.com/ckeditor/ckeditor5/pull/18966/files

Comment (martnpaneq) [reply .npmrc]: I am using the options that we agreed on during the `pnpm` meeting.

File: packages/ckeditor5-package-generator/lib/utils/choose-package-manager.js (hunk 1 above)
Comment (pomek) [reply packages/ckeditor5-package-generator/lib/utils/choose-package-manager.js:18]: We did it to detect the available package managers. Unfortunately, I am unsure whether it was necessary or if we used a _fancy_ switcher.

Comment (martnpaneq) [reply packages/ckeditor5-package-generator/lib/utils/choose-package-manager.js]: Removed this code in [1188077](https://github.com/ckeditor/ckeditor5-package-generator/pull/249/commits/11880771ea50faae55c30c026890145544fae1f9).

File: package.json [hunk 3]
```diff
@@ -56,8 +58,10 @@
     "release:publish-packages": "node scripts/release/publishpackages.js",
     "test": "vitest run --config vitest.config.js",
     "coverage": "vitest run --config vitest.config.js --coverage",
-    "clean": "npx rimraf --glob package-lock.json yarn.lock ./**/node_modules",
-    "reinstall": "yarn run clean && yarn install"
+    "clean": "pnpm dlx rimraf --glob ./**/node_modules",
```
[the remark this answers is not in the document]
Comment (martnpaneq) [reply package.json:61]: We agreed, that we are leaving the reinstall command. So what would you suggest?

[the remark this answers is not in the document]
Comment (przemyslaw-zan) [reply package.json:61]: We can leave it then.

**Review by psmyrek — COMMENTED** [review]
One file to update:

https://github.com/ckeditor/ckeditor5-package-generator/blob/6c2b76d87356dd1755cd55db2ab5462f3aebfb4e/.github/PULL_REQUEST_TEMPLATE.md?plain=1#L5

File: .circleci/config.yml [hunk 4]
```diff
@@ -175,6 +183,20 @@ jobs:
           name: Verify build (JavaScript, yarn, current and legacy installation methods)
           command: node scripts/ci/verify-build -l js -p yarn -m current-and-legacy -g GLOBAL_NAME
 
+  package_js_pnpm_both_install_methods:
+    docker:
+      - image: cimg/node:22.12.0-browsers
+    environment:
+      # To address the "error:0308010C:digital envelope routines::unsupported" issue.
+      NODE_OPTIONS: --openssl-legacy-provider
```
[the remark this answers is not in the document]
Comment (martnpaneq) [reply .circleci/config.yml]: I removed it in [7612762](https://github.com/ckeditor/ckeditor5-package-generator/pull/249/commits/761276244eadb56b387fb8ea97c00131bbc61ac0) and the build is fine, so I guess it's not needed any more.

File: .circleci/config.yml [hunk 5]
```diff
@@ -18,10 +18,18 @@ commands:
...
+            npm install -g pnpm@10.14.0 --prefix .pnpm
+            echo 'export PATH="$PWD/.pnpm/bin:$PATH"' >> $BASH_ENV
+      - run:
+          name: Verify pnpm version
+          command: pnpm --version
       - install_ssh_keys_command
       - run:
           name: Install dependencies
-          command: yarn install
+          command: pnpm install
```
[the remark this answers is not in the document]
Comment (martnpaneq) [reply .circleci/config.yml]: I used the first proposed solution: `npm install -g pnpm@10 --prefix .pnpm`.

Adressed in [ce454f8](https://github.com/ckeditor/ckeditor5-package-generator/pull/249/commits/ce454f8a87b02b67221d4556a8459e61d8b295ba).

File: .changelog/20250902144912_i_247_pnpm_as_package_manager.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+scope:
+  - ckeditor5-package-generator
+closes:
+  - 247
+---
+
+Added comprehensive support for `pnpm` alongside existing `npm` and `yarn` options in the CKEditor 5 package generator.
+
+Features added:
+- New `--use-pnpm` CLI flag for specifying pnpm as package manager
```
[the remark this answers is not in the document]
Comment (martnpaneq) [reply .changelog/20250902144912_i_247_pnpm_as_package_manager.md]: Fixed in [ce454f8](https://github.com/ckeditor/ckeditor5-package-generator/pull/249/commits/ce454f8a87b02b67221d4556a8459e61d8b295ba).

File: packages/ckeditor5-package-generator/lib/utils/choose-package-manager.js (hunk 1 above)
Comment (martnpaneq) [reply packages/ckeditor5-package-generator/lib/utils/choose-package-manager.js:18]: I reverted the removal of detection if package manager is installed in [279e013](https://github.com/ckeditor/ckeditor5-package-generator/pull/249/commits/279e01341bc587837a119c4298d8bf0a1ff0d9b0).

Comment (martnpaneq) [reply packages/ckeditor5-package-generator/lib/utils/choose-package-manager.js:18]: We need to think now, whether we should display package managers in the prompt if they are not installed @pomek .

File: packages/ckeditor5-package-generator/lib/utils/choose-package-manager.js [hunk 7]
```diff
@@ -12,9 +14,18 @@ import inquirer from 'inquirer';
...
-	if ( selected > 1 ) {
+	if ( useYarn && !yarnInstalled ) {
+		throw new Error( 'Detected --use-yarn option but yarn is not installed.' );
+	}
+
+	if ( usePnpm && !pnpmInstalled ) {
+		throw new Error( 'Detected --use-pnpm option but pnpm is not installed.' );
+	}
+
+	if ( useNpm && usePnpm && useYarn || useNpm && useYarn || useNpm && usePnpm || usePnpm && useYarn ) {
```
[the remark this answers is not in the document]
Comment (martnpaneq) [reply packages/ckeditor5-package-generator/lib/utils/choose-package-manager.js]: Readded in [4a467a2](https://github.com/ckeditor/ckeditor5-package-generator/pull/249/commits/4a467a26898aa568ee163a336ca8fa178cd2b9b7).
